File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit_objc/FIAPRequestHandler.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "./include/in_app_purchase_storekit_objc/FIAPRequestHandler.h"
#import <StoreKit/StoreKit.h>

#pragma mark - Main Handler

@interface FIAPRequestHandler () <SKProductsRequestDelegate>

@property(nonatomic, copy) ProductRequestCompletion completion;
@property(nonatomic, strong) SKRequest *request;

@end

@implementation FIAPRequestHandler

- (instancetype)initWithRequest:(SKRequest *)request {
  self = [super init];
  if (self) {
    self.request = request;
    request.delegate = self;
  }
  return self;
}

- (void)startProductRequestWithCompletionHandler:(ProductRequestCompletion)completion {
  self.completion = completion;
  [self.request start];
}

- (void)productsRequest:(SKProductsRequest *)request
     didReceiveResponse:(SKProductsResponse *)response {
  if (self.completion) {
    self.completion(response, nil);
    // set the completion to nil here so self.completion won't be triggered again in
    // requestDidFinish for SKProductRequest.
    self.completion = nil;
  }
}

- (void)requestDidFinish:(SKRequest *)request {
  if (self.completion) {
    self.completion(nil, nil);
  }
}

- (void)request:(SKRequest *)request didFailWithError:(NSError *)error {
  if (self.completion) {
    self.completion(nil, error);
  }
}

@end

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit/FIAPRequestHandler.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import StoreKit

public class FIAPRequestHandler: NSObject, FLTRequestHandlerProtocol {
  private var completion: ((SKProductsResponse?, Error?) -> Void)?
  private let request: SKRequest

  public init(request: SKRequest) {
    self.request = request
    super.init()
    request.delegate = self
  }

  public func startProductRequest(
    completionHandler: @escaping (SKProductsResponse?, Error?) -> Void
  ) {
    completion = completionHandler
    request.start()
  }
}

extension FIAPRequestHandler: SKProductsRequestDelegate {
  public func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse)
  {
    if let completion = completion {
      completion(response, nil)
      // set the completion to nil here so completion won't be triggered again in
      // requestDidFinish for SKProductRequest.
      self.completion = nil
    }
  }
}

extension FIAPRequestHandler: SKRequestDelegate {
  public func requestDidFinish(_ request: SKRequest) {
    completion?(nil, nil)
  }

  public func request(_ request: SKRequest, didFailWithError error: Error) {
    completion?(nil, error)
  }
}
